modules/edge-ren/src/lib.rs

```diff
@@ -0,0 +1,464 @@
+}
+
+type RenTokenInfoType<T> = RenTokenInfo<<T as Config>::RenVMTokenIdType, TokenIdOf<T>>;
+
+decl_storage! {
+	trait Store for Module<T: Config> as Template {
+		/// Signature blacklist. This is required to prevent double claim. Unbounded
+		Signatures get(fn signatures): map hasher(opaque_twox_256) EcdsaSignature => Option<()>;
+		/// Record burn event details
+		BurnEvents get(fn burn_events): map hasher(twox_64_concat) u32 => Option<(<T as Config>::RenVMTokenIdType, T::BlockNumber, DestAddress, BalanceOf<T>)>;
+		/// Next burn event ID
+		NextBurnEventId get(fn next_burn_event_id): u32;
+		/// Registry of all active tokens
+		RenTokenRegistry get(fn ren_token_registry): map hasher(blake2_128_concat) <T as Config>::RenVMTokenIdType => Option<RenTokenInfoType<T>>;
+	}
+}
+
+decl_event!(
+	pub enum Event<T> where
+		<T as frame_system::Config>::AccountId,
+		<T as Config>::RenVMTokenIdType,
+		Balance = BalanceOf<T>
+	{
+		RenTokenAdded(RenVMTokenIdType),
+
+		RenTokenUpdated(RenVMTokenIdType),
+
+		RenTokenDeleted(RenVMTokenIdType),
+
+		RenTokenSpent(RenVMTokenIdType, Balance),
+
+		RenTokenMinted(RenVMTokenIdType, AccountId, Balance),
+
+		RenTokenBurnt(RenVMTokenIdType, AccountId, DestAddress, Balance),
+
+	}
+);
+
+decl_error! {
+	pub enum Error for Module<T: Config> {
+		/// The mint signature is invalid.
+		InvalidMintSignature,
+		/// Burn ID overflow.
+		BurnIdOverflow,
+		/// RenTokenAlready Exists
+		RenTokenAlreadyExists,
+		/// No token with this ren_token_id found
+		RenTokenNotFound,
+		/// Token name exceeds length limit
+		RenTokenNameLengthLimitExceeded,
+		/// The to address provided to the burn function failed validation corresponding to the token's name
+		InvalidBurnToAddress,
+		/// An operation that was not expected to fail failed
+		UnexpectedError,
+		/// Minting this token has been disabled
+		RenTokenMintDisabled,
+		/// Burning this token has been disabled
+		RenTokenBurnDisabled
+
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Config> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		/// Instantiates the token in the Registry
+		/// Must be done AFTER instantiating the asset in Assets
+		/// max length of the token name implemented
+		#[weight = 10_000]
+		fn add_ren_token(
+			origin,
+			#[compact] _ren_token_id: T::RenVMTokenIdType,
+			_ren_token_asset_id: TokenIdOf<T>,
+			_ren_token_name: Vec<u8>,
+			_ren_token_renvm_id: [u8; 32],
+			_ren_token_pub_key: [u8; 20],
+			_ren_token_mint_enabled: bool,
+			_ren_token_burn_enabled: bool,
+			_ren_token_mint_fee: u32,
+			_ren_token_burn_fee: u32,
+		) -> DispatchResult
+		{
+			T::ControllerOrigin::ensure_origin(origin)?;
+
+			ensure!(!<RenTokenRegistry<T>>::contains_key(&_ren_token_id), Error::<T>::RenTokenAlreadyExists);
+			ensure!(_ren_token_name.len()<=NAME_MAX_LENGTH.into(), Error::<T>::RenTokenNameLengthLimitExceeded);
+
+			// TODO Need to check if asset exists in Assets
```

Comment:
Additionally, we should attempt to create a new asset if adding a new ren token directly here.

Reply:
1. Since we are using traits for the most part, do we really want to tightly couple (extend) the pallet_assets just to query the storage? Perhaps we would like to create a new trait that delivers the necessary information of a multi-currency fungible asset system, which has get functions such as: does the currency with the assetId exist?, what is the minimum balance for that asset?,  etc... Or perhaps integrate the ones we need into the currently existing ManageableAsset trait for the time being.
2. Since there is no trait to create an asset nor are the create functions public, I do not know how to approach creating new asset from the edge-ren pallet. Perhaps an ideal solution would be to enable unrestricted crud of assets from the ManageableAsset trait.

Reply:
Also, It seems to me that without using traits (atleast for create and destroy) we would need to do benchmarking in its own crate.

Reply:
Creating new asset from the edge-ren pallet (even though the functions are private), seems to be possible via "pallet_assets::Call::<T>::".
But as the runtime type bounds don't seem to be determined while compiling the pallet, generic ascii or hex data cannot be used as input to functions of the pallet-assets accessed this way (Note this is possible in testing where the the runtime is compiled with the pallet's test code).  The inputs of the function need to be the associated type of the pallet. This results in the scenario discussed in https://github.com/hicommonwealth/edgeware-node/pull/272#discussion_r549331151, which is an issue when using traits for the mint, burn and transfer calls.